src/runtime-rs/crates/resource/src/cpu_mem/cpu.rs

```diff
@@ -0,0 +1,187 @@
+                ResourceUpdateOp::Update => {
+                    let resource = resources.insert(cid.to_owned(), container_resource.clone());
+                    if let Some(old_container_resource) = resource {
+                        // the priority of cpu-quota is higher than cpuset when determine the number of vcpus.
+                        // we should better ignore the resource update when update cpu only by cpuset if cpu-quota
+                        // has been set previously.
+                        if old_container_resource.quota() > 0 && container_resource.quota() < 0 {
+                            resources.insert(cid.to_owned(), old_container_resource);
+                        }
+                    }
+                }
+                ResourceUpdateOp::Del => {
+                    resources.remove(cid);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    // calculates the total required vcpus by adding each container's requirements within the pod
+    async fn calc_cpu_resources(&self) -> Result<u32> {
+        let mut total_vcpu = 0;
+        let mut cpuset_vcpu: HashSet<u32> = HashSet::new();
+
+        let resources = self.container_cpu_resources.read().await;
+        for (_, cpu_resource) in resources.iter() {
+            let vcpu = cpu_resource.get_vcpus().unwrap_or(0) as u32;
+            cpuset_vcpu.extend(cpu_resource.cpuset().iter());
+            total_vcpu += vcpu;
+        }
+
+        // contrained only by cpuset
+        if total_vcpu == 0 && !cpuset_vcpu.is_empty() {
+            info!(sl!(), "(from cpuset)get vcpus # {:?}", cpuset_vcpu);
+            return Ok(cpuset_vcpu.len() as u32);
+        }
+
+        info!(
+            sl!(),
+            "(from cfs_quota&cfs_period)get vcpus count {}", total_vcpu
+        );
+        Ok(total_vcpu)
+    }
+
+    // do hotplug the vcpu
+    async fn do_update_cpu_resources(
+        &self,
+        new_vcpus: u32,
+        op: ResourceUpdateOp,
+        hypervisor: &dyn Hypervisor,
+        agent: &dyn Agent,
+    ) -> Result<u32> {
+        let old_vcpus = self.current_vcpu().await;
+
+        // when adding vcpus, ignore old_vcpus > new_vcpus
+        // when deleting vcpus, ignore old_vcpus < new_vcpus
+        if (op == ResourceUpdateOp::Add && old_vcpus > new_vcpus)
+            || (op == ResourceUpdateOp::Del && old_vcpus < new_vcpus)
+        {
+            return Ok(old_vcpus);
+        }
+
+        // do not reduce computing power
+        // the number of vcpus would not be lower than the default size
+        let new_vcpus = cmp::max(new_vcpus, self.default_vcpu);
+
+        let (old, new) = hypervisor
+            .resize_vcpu(old_vcpus, new_vcpus)
+            .await
+            .context("resize vcpus")?;
+
+        if old < new {
+            let add = new - old;
+            info!(sl!(), "request to onlineCpuMem with {:?} cpus", add);
+
+            agent
+                .online_cpu_mem(OnlineCPUMemRequest {
```

Comment:
Why this request is called `OnlineCPUMemRequest` ? Do we want to interwine cpu & memory online into one single request? What's the reason for this mixed design?

Reply:
This request is inherited from runtime-go.